module/VuFindSearch/src/VuFindSearch/Backend/Solr/Connector.php

```diff
@@ -404,26 +423,60 @@ protected function forceToBackendException($ex)
      * @param string   $method    HTTP method to use
      * @param string   $urlSuffix Suffix to append to all URLs tried
      * @param callable $callback  Callback to configure client (null for none)
+     * @param bool     $cacheable Whether the request is cacheable
      *
      * @return string Response body
      *
      * @throws RemoteErrorException  SOLR signaled a server error (HTTP 5xx)
      * @throws RequestErrorException SOLR signaled a client error (HTTP 4xx)
      */
-    protected function trySolrUrls($method, $urlSuffix, $callback = null)
-    {
+    protected function trySolrUrls($method, $urlSuffix, $callback = null,
+        bool $cacheable = false
+    ) {
         // This exception should never get thrown; it's just a safety in case
         // something unanticipated occurs.
         $exception = new \Exception('Unexpected exception.');
 
         // Loop through all base URLs and try them in turn until one works.
+        $cacheKey = null;
         foreach ((array)$this->url as $base) {
             $client = $this->createClient($base . $urlSuffix, $method);
             if (is_callable($callback)) {
                 $callback($client);
             }
+            // Always create the cache key from the first server, and only after any
+            // callback has been called above.
+            if ($cacheable && $this->cache && null === $cacheKey) {
+                $cacheKey = md5($client->getRequest()->toString());
+                try {
+                    if ($result = $this->cache->getItem($cacheKey)) {
+                        $this->debug('Returning cached results');
+                        return $result;
+                    }
+                } catch (\Exception $ex) {
+                    $this->logWarning('Cache getItem failed: ' . $ex->getMessage());
+                }
+            }
             try {
-                return $this->send($client);
+                $result = $this->send($client);
+
+                if ($cacheKey) {
+                    try {
+                        $this->cache->setItem($cacheKey, $result);
+                    } catch (\Exception $ex) {
+                        if ($ex->getMessage() === 'ITEM TOO BIG') {
```

Comment:
Is there a more reliable way to detect this condition? Is a more specific exception thrown, or is there a code that might be more stable over time than the textual message?

Reply:
Well, it's actually Laminas\Cache\Exception\RuntimeException, but it's still just the string from Memcached object's getResultMessage(). Memcached has a getResultCode method as well, but Laminas' adapter doesn't use it.

Reply:
That's frustrating, but it sounds like there's not much we can do about it. Maybe the best solution for now is simply to add a comment explaining why this magic string is here, so we can revisit in the future if things change, and so people will better understand why this is here.

Reply:
I added a comment and revised the exception handling a bit in anticipation of codes being available at some point. There's the additional complication that any error codes are not translated, so any checks should be adapter-specific. Nevertheless, I posted https://github.com/laminas/laminas-cache-storage-adapter-memcached/pull/6 to add error codes to Memcached exceptions since other adapters already include them.

Reply:
Thanks! I see that they're asking for test coverage on your Laminas PR, which is a good sign. :-)